pkg/pillar/cmd/zedrouter/acl.go

```diff
@@ -770,141 +800,235 @@ func diffIpsets(newIpsets, oldIpsets []string) ([]string, []string, bool) {
 	}
-	if !isMgmt {
-		// Remove mangle rules for IPv6 packets added above
-		// XXX error checks?
-		iptables.Ip6tableCmd("-t", "mangle", "-D", "PREROUTING", "-i", bridgeName,
-			"-p", "tcp", "-j", "CHECKSUM", "--checksum-fill")
-		iptables.Ip6tableCmd("-t", "mangle", "-D", "PREROUTING", "-i", bridgeName,
-			"-p", "udp", "-j", "CHECKSUM", "--checksum-fill")
+	return true
+}
+
+// utility routines for IpTables Rules
+func executeIPTablesRule(operation string, rule types.IPTablesRule) error {
+	var err error
+	ruleStr := []string{}
+	if rule.Table != "" {
+		ruleStr = append(ruleStr, "-t")
+		ruleStr = append(ruleStr, rule.Table)
+	}
+	ruleStr = append(ruleStr, operation)
+	ruleStr = append(ruleStr, rule.Chain)
+	ruleStr = append(ruleStr, rule.Prefix...)
+	ruleStr = append(ruleStr, rule.Rule...)
+	if rule.IPVer == 4 {
+		err = iptables.IptableCmd(ruleStr...)
+	} else if rule.IPVer == 6 {
+		err = iptables.Ip6tableCmd(ruleStr...)
+	} else {
+		errStr := fmt.Sprintf("ACL: Unknown IP version %d", rule.IPVer)
+		err = errors.New(errStr)
 	}
-	// XXX see above
-	if false && ipVer == 6 && !isMgmt {
-		// Manually delete the manual add above
-		// XXX error checks?
-		iptables.Ip6tableCmd("-D", "FORWARD", "-i", bridgeName, "-o", "dbo1x0",
-			"-j", "DROP")
+	return err
+}
+
+// handle network instance level ACL rules
+// Network Instance Level ACL rule handling routines
+func handleNetworkInstanceACLConfiglet(op string, aclArgs types.AppNetworkACLArgs) error {
+
+	log.Infof("bridge(%s, %v) iptables op: %v\n", aclArgs.BridgeName, aclArgs.BridgeIP, op)
+	rulesList := networkInstanceBridgeRules(aclArgs)
+	for _, rule := range rulesList {
+		if err := executeIPTablesRule(op, rule); err != nil {
+			return err
+		}
 	}
 	return nil
 }
+
+func networkInstanceBridgeRules(aclArgs types.AppNetworkACLArgs) types.IPTablesRuleList {
+	var rulesList types.IPTablesRuleList
+	var aclRule1, aclRule2 types.IPTablesRule
+
+	// not for dom0
+	if aclArgs.IsMgmt {
+		return rulesList
+	}
+	aclArgs.IPVer = determineIPVer(aclArgs.IsMgmt, aclArgs.BridgeIP)
+	// two rules for ipv4
+	aclRule1.IPVer = 4
+	aclRule1.Table = "mangle"
+	aclRule1.Chain = "PREROUTING"
+	aclRule1.Rule = []string{"-i", aclArgs.BridgeName, "-p", "tcp",
+		"-j", "CHECKSUM", "--checksum-fill"}
+	aclRule2.IPVer = 4
+	aclRule2.Table = "mangle"
+	aclRule2.Chain = "PREROUTING"
+	aclRule2.Rule = []string{"-i", aclArgs.BridgeName, "-p", "udp",
+		"-j", "CHECKSUM", "--checksum-fill"}
+	rulesList = append(rulesList, aclRule1, aclRule2)
+
+	// two rules for ipv6
+	aclRule1.IPVer = 6
+	aclRule1.Table = "mangle"
+	aclRule1.Chain = "PREROUTING"
+	aclRule1.Rule = []string{"-i", aclArgs.BridgeName, "-p", "tcp",
+		"-j", "CHECKSUM", "--checksum-fill"}
+	aclRule2.IPVer = 6
+	aclRule2.Table = "mangle"
+	aclRule2.Chain = "PREROUTING"
+	aclRule2.Rule = []string{"-i", aclArgs.BridgeName, "-p", "udp",
+		"-j", "CHECKSUM", "--checksum-fill"}
+	rulesList = append(rulesList, aclRule1, aclRule2)
+
+	aclRule1.IPVer = aclArgs.IPVer
```

Comment:
You have these forward to rules in aclDroprules and here. But the forward from drop rules are only on aclDroprules.

I think you can put all of them here and delete aclDroprules() completely.
That would also ensure (with the -I in reverse order) that the drop rules are always at the end as long as the network instance adds its rules before any app instances insert there rules.
However, you need to check that we don't have any modify code for the network instance which deletes and re-adds the ACLs.

Reply:
Or you can keep the aclDropRules function but only call it from the networkInstanceBridgeRules() function.